### PR TITLE
[GraphBolt] Refactor `FeatureFetcher` to use `read_async`.

### PR DESCRIPTION
**Pull request by mfbalin**
## Description
Depended on #7558. Tested with our examples. The CI should pass now.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/feature_fetcher.py [hunk 1]
```diff
@@ -49,14 +67,95 @@ def __init__(
...
+                    for type_name, feature_names in edge_feature_keys.items()
+                    for feature_name in feature_names
+                ]
+            elif edge_feature_keys is not None:
+                edge_feature_key_list = [
+                    ("edge", None, feature_name)
+                    for feature_name in edge_feature_keys
+                ]
+            else:
+                edge_feature_key_list = []
```
[the remark this answers is not in the document]
Comment (mfbalin) [reply python/dgl/graphbolt/feature_fetcher.py:103]: This is how the existing code was.

[the remark this answers is not in the document]
Comment (mfbalin) [reply python/dgl/graphbolt/feature_fetcher.py:103]: Yeah, I missed the opportunity. Will refactor.

File: python/dgl/graphbolt/feature_fetcher.py [hunk 2]
```diff
@@ -49,14 +67,95 @@ def __init__(
...
+            for feature_key_list in [
+                node_feature_key_list,
+                edge_feature_key_list,
+            ]:
+                for feature_key in feature_key_list:
+                    if feature_key not in feature_store:
+                        continue
+                    for device_str in ["cpu", "cuda"]:
+                        try:
+                            max_val = max(
```
[the remark this answers is not in the document]
Comment (mfbalin) [reply python/dgl/graphbolt/feature_fetcher.py:113]: Yes. The only case that does not work is if tensor is stored on GPU already. I am coming up with a fix for it now.

[the remark this answers is not in the document]
Comment (mfbalin) [reply python/dgl/graphbolt/feature_fetcher.py:101]: Fixed it with the latest commit.

**Pull request merged**